src/video.js

```diff
@@ -64,3 +69,59 @@ export const checkVideoBidSetup = hook('sync', function(bid, adUnit, videoMediaT
 
   return true;
 }, 'checkVideoBidSetup');
+
+export function registerVastTrackers(moduleType, moduleName, trackerFn) {
+  if (typeof trackerFn == 'function') { vastTrackers.push({'moduleType': moduleType, 'moduleName': moduleName, 'trackerFn': trackerFn}); }
+}
+
+export function insertVastTrackers(trackers, vastXml) {
+  const doc = new DOMParser().parseFromString(vastXml, 'text/xml');
+  const wrappers = doc.querySelectorAll('VAST Ad Wrapper, VAST Ad InLine');
+  try {
+    if (wrappers.length) {
+      wrappers.forEach(wrapper => {
+        if (trackers.get('impression')) {
+          trackers.get('impression').forEach(trackingUrl => {
+            const impression = doc.createElement('Impression');
+            impression.appendChild(doc.createCDATASection(trackingUrl));
+            wrapper.appendChild(impression)
+          });
+        }
+      });
+      vastXml = new XMLSerializer().serializeToString(doc);
+    }
+  } catch (error) {
+    logError('an error happened trying to insert trackers in vastXml');
+  }
+  return vastXml;
+}
+
+export function getVastTrackers(bid) {
+  let trackers = [];
+  vastTrackers.filter(
+    ({moduleType, moduleName, trackerFn}) => isActivityAllowed(ACTIVITY_REPORT_ANALYTICS, activityParams(moduleType, moduleName))
+  ).forEach(({trackerFn}) => {
+    let trackersToAdd = trackerFn(bid);
+    trackersToAdd.forEach(trackerToAdd => {
+      if (validVastTracker(trackers, trackerToAdd)) { trackers.push(trackerToAdd); }
+    });
+  });
+  const trackersMap = trackersToMap(trackers);
+  return (trackersMap.size ? trackersMap : null);
+};
+
+function validVastTracker(trackers, trackerToAdd) {
+  if (!trackerToAdd.hasOwnProperty('event') || !trackerToAdd.hasOwnProperty('url')) { return false; }
+  trackers.forEach(tracker => {
+    if (tracker['event'] == trackerToAdd['event'] && tracker['url'] == trackerToAdd['url']) { return false; }
```

Comment:
note: this `return false` won't return from `isValidVastTracker` , because in `forEach`, the argument is a function. So your return statement is returning from the callback in the loop. So it doesn't do anything.

I suggest in this case using a traditional for loop

Reply:
Indeed. Actually this step was to avoid to add multiple times the exact same url for a same tracking event. While I don't really see the point in adding duplicates tracking calls, I am not sure that's really something we should be checking for but more a responsability from those adding trackers. I've removed it.